python/sbp/client/util/fftmonitor.py

```diff
@@ -0,0 +1,196 @@
+except ImportError:
+    from sbp.piksi import SBP_MSG_SPECAN, MsgSpecan
+import numpy as np
+
+class GPSTimeSecCmp(GPSTimeSec):
+    def __init__(self, payload=None, **kwargs):
+        super(GPSTimeSecCmp, self).__init__(payload=payload, **kwargs)
+
+    def __cmp__(self, other):
+        if other is None:
+            return 1
+        elif self.wn < other.wn:
+            return -1
+        elif self.wn > other.wn:
+            return 1
+        else:
+            if self.tow < other.tow:
+                return -1
+            elif self.tow > other.tow:
+                return 1
+            else:
+                return 0
+
+    def __eq__(self, other):
+        if not isinstance(other, GPSTimeSecCmp):
+            return False
+        return (self.wn == other.wn) and (self.tow == other.tow)
+
+    def __repr__(self):
+        return '(wn: {0}, tow: {1})'.format(self.wn, self.tow)
+
+    def __hash__(self):
+        return (self.wn, self.tow).__hash__()
+
+
+class FFTMonitor(object):
+    CHANNELS = [1, 2, 3, 4]
+
+    def __init__(self):
+        self.mkvalue = lambda: {
+            'frequencies': np.array([]),
+            'amplitudes': np.array([])}
+        self.ffts = {k: [] for k in FFTMonitor.CHANNELS}
+        self.incomplete_ffts = {
+            ch: defaultdict(self.mkvalue) for ch in FFTMonitor.CHANNELS}
+        self.enabled = {i: None for i in FFTMonitor.CHANNELS}
+
+    def get_frequencies(self, sbp_msg):
+        ref = sbp_msg.freq_ref
+        step = sbp_msg.freq_step
+        n = len(sbp_msg.amplitude_value)
+        return np.array([ref + step * i for i in range(n)])
+
+    def get_amplitudes(self, sbp_msg):
+        ref = sbp_msg.amplitude_ref
+        unit = sbp_msg.amplitude_unit
+        values = sbp_msg.amplitude_value
+        return np.array([ref + unit * value for value in values])
+
+    def capture_fft(self, sbp_msg, **metadata):
+
+        # print "ch:%d, freq_ref: %f, freq_step: %f, amp_ref:%f, num_values:%d, wn:%d, tow:%d" \
+        #         % (sbp_msg.channel_tag, sbp_msg.freq_ref, sbp_msg.freq_step, sbp_msg.amplitude_ref, len(sbp_msg.amplitude_value), sbp_msg.t.wn, sbp_msg.t.tow)
+
+        channel = sbp_msg.channel_tag
+        if not self.enabled[channel]:
+            return
+
+        frequencies = self.get_frequencies(sbp_msg)
+        amplitudes = self.get_amplitudes(sbp_msg)
+        timestamp = GPSTimeSecCmp(tow=sbp_msg.t.tow, wn=sbp_msg.t.wn)
+
+        self.incomplete_ffts[channel][timestamp]['frequencies'] = np.append(
+            self.incomplete_ffts[channel][timestamp]['frequencies'],
+            frequencies,
+            axis=0)
+
+        self.incomplete_ffts[channel][timestamp]['amplitudes'] = np.append(
+            self.incomplete_ffts[channel][timestamp]['amplitudes'],
+            amplitudes,
+            axis=0)
+
+        if len(self.incomplete_ffts[channel][timestamp]['frequencies']) == 512:
+            current_fft = self.incomplete_ffts[channel].pop(timestamp)
```

Comment:
What's the expected behavior if you miss a message?

Reply:
Ignore it and wait for the next one.